ocaml/xapi/xapi_pool.ml

```diff
@@ -118,6 +112,71 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
             )
         )
   in
+  let one_ip_configured_on_joining_cluster_network () =
+    match Client.Cluster_host.get_all ~rpc ~session_id with
+    | [] ->
+        ()
+    | ch :: _ -> (
+      match Client.Cluster_host.get_PIF ~rpc ~session_id ~self:ch with
+      | pif when pif = Ref.null ->
+          ()
+      | pif -> (
+        match Client.PIF.get_VLAN ~rpc ~session_id ~self:pif with
+        | vlan when vlan > 0L ->
+            error "Cannot join pool whose clustering is enabled on VLAN network" ;
+            raise
+              (Api_errors.Server_error
+                 ( Api_errors
+                   .pool_joining_pool_cannot_enable_clustering_on_vlan_network
+                 , [Int64.to_string vlan]
+                 )
+              )
+        | 0L | _ -> (
+            let bonds =
+              Client.PIF.get_bond_master_of ~rpc ~session_id ~self:pif
+            in
+            let expectedDevicesWithIp =
+              match bonds with
+              | [] ->
+                  [Client.PIF.get_device ~rpc ~session_id ~self:pif]
+              | _ ->
+                  bonds
+                  |> List.map (fun bond ->
```

Comment:
`List.concat_map`?

Reply:
And how about the following?
```
let clustering_devices_in_pool =
  (match Client.PIF.get_bond_master_of ~rpc ~session_id ~self:pif with
  | [] ->
      [pif]
  | bonds ->
      List.concat_map
        (fun bond -> Client.Bond.get_slaves ~rpc ~session_id ~self:bond) bonds
  )
  |> List.map (fun self -> Client.PIF.get_device ~rpc ~session_id ~self)
in
Db.Host.get_PIFs ~__context ~self:(Helpers.get_localhost ~__context)
|> List.map (fun self -> (p, Db.PIF.get_device ~__context ~self)
|> List.filter (fun _, d -> List.mem d clustering_devices_in_pool)
|> List.filter (fun p, _ -> Db.PIF.get_IP ~__context ~self:p <> "")
| [] ->
    ()
| _::_ ->
    ...
```

Reply:
I would agree with @minglumlu to reduce the nesting by introducing some helper functions in between.